gix-dir/src/walk/classify.rs

```diff
@@ -169,6 +169,7 @@ pub fn path(
                         .map(|platform| platform.excluded_kind())
                 })
                 .map_err(Error::ExcludesAccess)?
+                .filter(|_| filename_start_idx > 0)
```

Comment:
This line and the line below constitute the fix, it's quite neat.

The idea is that previously, the classifier didn't distinguish between the worktree root (and the contained repository) and sub-repositories. For the latter, the current classification is very much intentional, but the root should never ever be counted as ignored, and thus removable.

The commit [with the fix](https://github.com/Byron/gitoxide/pull/1462/commits/1e92d1ed3f23e9de84203252d64e701e2c7fb941) also shows how the classification changes due to the fix, which is all it takes to prevent `gix clean` to go off the rail.

When looking for weaknesses, I also recommend using the pathspec, as it is quite integrated to the algorithm to have some (possibly unexpected) influence.